lib/mongo/collection/view/change_stream.rb

```diff
@@ -0,0 +1,138 @@
+# limitations under the License.
+
+require 'mongo/collection/view/change_stream/retryable'
+
+module Mongo
+  class Collection
+    class View
+
+      # Provides behaviour around a `$changeStream` pipeline stage in the
+      # aggregation framework. Specifying this stage allows users to request that
+      # notifications are sent for all changes to a particular collection or database.
+      #
+      # @note Only available in server versions 3.6 and higher.
+      #
+      # @since 2.5.0
+      class ChangeStream < Aggregation
+        include Retryable
+
+        # @return [ String ] The fullDocument option default value.
+        #
+        # @since 2.5.0
+        FULL_DOCUMENT_DEFAULT = 'default'.freeze
+
+        # @return [ BSON::Document ] The change stream options.
+        #
+        # @since 2.5.0
+        attr_reader :options
+
+        # Initialize the change stream for the provided collection view, pipeline
+        # and options.
+        #
+        # @example Create the new change stream view.
+        #   ChangeStream.new(view, pipeline, options)
+        #
+        # @param [ Collection::View ] view The collection view.
+        # @param [ Array<Hash> ] pipeline The pipeline of operators to filter the change notifications.
+        # @param [ Hash ] opts The change stream options.
+        #
+        # @option options [ String ] :full_document Allowed values: ‘default’, ‘updateLookup’. Defaults to ‘default’.
+        #   When set to ‘updateLookup’, the change notification for partial updates will include both a delta
+        #   describing the changes to the document, as well as a copy of the entire document that was changed
+        #   from some time after the change occurred.
+        # @option options [ BSON::Document, Hash ] :resume_after Specifies the logical starting point for the
+        #   new change stream.
+        # @option options [ Integer ] :max_await_time_ms The maximum amount of time for the server to wait
+        #   on new documents to satisfy a change stream query.
+        # @option options [ Integer ] :batch_size The number of documents to return per batch.
+        # @option options [ BSON::Document, Hash ] :collation The collation to use.
+        #
+        # @since 2.5.0
+        def initialize(view, pipeline, options = {})
+          @view = view
+          @change_stream_filters = pipeline
+          @options = options.freeze
+          @resume_token = @options[:resume_after]
+          create_cursor!
+        end
+
+        # Iterate through documents returned by the change stream.
+        #
+        # @example Iterate through the stream of documents.
+        #   stream.each do |document|
+        #     p document
+        #   end
+        #
+        # @return [ Enumerator ] The enumerator.
+        #
+        # @since 2.5.0
+        #
+        # @yieldparam [ BSON::Document ] Each change stream document.
+        def each
```

Comment:
I think there might be a few spec violations in this method. First, the spec says:
> A driver SHOULD attempt to kill the cursor on the server on which the cursor is opened during the resume process, and MUST NOT attempt to kill the cursor on any other server.

Does `close` ensure that the kill cursor request goes to the original server?

Second, `create_cursor` retrys once and this method retrys as well. That means it will retry on consecutive network errors. The spec says to only retry on the first network error. Any errors that happen during the retry should be fatal. I think instead of `(@cursor || create_cursor!)` it should be something like `(@cursor || create_cursor_no_retry!)`.

Reply:
> Does close ensure that the kill cursor request goes to the original server?

Yes. A cursor has a reference to the server (``@server`` variable). The kill cursors method called on the cursor in line 105, [sends the operation to that server](https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/cursor.rb#L201).

> Second, create_cursor retrys once and this method retrys as well. That means it will retry on consecutive network errors. The spec says to only retry on the first network error. Any errors that happen during the retry should be fatal. I think instead of (@cursor || create_cursor!) it should be something like (@cursor || create_cursor_no_retry!). Do you agree or am I misunderstanding the code?

Yes, you're right. I shouldn't be retrying in the create_cursor method. I altered the logic [a few commits ago](https://github.com/mongodb/mongo-ruby-driver/pull/888/commits/5cb322d8bf2f6a26a217aadc5ea7b98457007bf1). I believe it was correct before so I'll change it back.